spanner/ddl/ast.go

```diff
@@ -18,113 +18,65 @@
 //
 // Definitions are from
 // https://cloud.google.com/spanner/docs/data-definition-language.
-// Before defining each type, we give the snippet from this definition.
-//
-// We use go interface types to preserve the structural constraints of
-// Spanner DDL. For example, suppose ScalarType could be either
-// an INT or a STRING with a length specifier:
-//    ScalarType := INT | STRING( length )
-// then we use a go interface type to encode ScalarType:
-//    type ScalarType interface {
-//        PrintScalarType() string  // To print ScalarTypes types.
-//    }
-// and struct types for each case:
-//    type Int struct { }
-//    type String struct { length int64 }
-// and finally, we define functions:
-//    func (i Int) PrintScalarType() { ... }
-//    func (s String) PrintScalarType() { .. }
-//
-// The net result is that the only way to build a ScalarType
-// is using Int or String.
 package ddl
 
 import (
 	"fmt"
 	"strings"
 )
 
-// Length encodes the following Spanner DDL definition:
-//     length:
-//        { int64_value | MAX }
-type Length interface {
-	PrintLength() string
-}
-
-// Int64Length wraps an integer length specifier.
-type Int64Length struct{ Value int64 }
-
-// MaxLength represents "MAX".
-type MaxLength struct{}
-
-// Interface validation for Length.
-var _ = []Length{Int64Length{}, MaxLength{}}
-
-// PrintLength unparses Int64Length.
-func (i Int64Length) PrintLength() string { return fmt.Sprintf("%d", i.Value) }
+const (
+	// Bool represent BOOL type.
+	Bool string = "BOOL"
+	// Bytes represent BYTES type.
+	Bytes string = "BYTES"
+	// Date represent DATE type.
+	Date string = "DATE"
+	// Float64 represent FLOAT64 type.
+	Float64 string = "FLOAT64"
+	// Int64 represent INT64 type.
+	Int64 string = "INT64"
+	// String represent STRING type.
+	String string = "STRING"
+	// Timestamp represent TIMESTAMP type.
+	Timestamp string = "TIMESTAMP"
+	// MaxLength represent "MAX" length of column type.
+	MaxLength string = "MAX"
+)
 
-// PrintLength unparses MaxLength.
-func (m MaxLength) PrintLength() string { return fmt.Sprintf("MAX") }
+// PrintLength unparses length of a Type.
+func PrintLength(m interface{}) string {
+	return fmt.Sprintf("%v", m)
+}
 
-// ScalarType encodes the following DDL definition:
-//     scalar_type:
+// Type represents the type of a column.
+//     type:
 //        { BOOL | INT64 | FLOAT64 | STRING( length ) | BYTES( length ) | DATE | TIMESTAMP }
-type ScalarType interface {
-	PrintScalarType() string
+type Type struct {
+	Name string
+	// Len encodes the following Spanner DDL definition:
+	//     length:
+	//        { int64_value | MAX }
+	Len interface{}
+	// IsArray represents if Type is an array_type or not
+	// When false, column has type T; when true, it is an array of type T.
+	IsArray bool
```

Comment:
Moved IsArray from ColumnDef to Type. Also, removed all declarations of ScalarType.